### DISCUSSION TITLE
Re-introducing the NDDataArray?

### DISCUSSION THREAD

**Discussion by registerrier** [gammapy/gammapy#4461] [Ideas]
The new data model of CTAO relies on a N-dimension data array concept to represent tabulated functions. It is very similar to what we use as a base class in gammapy.irf, but less specialized.

We used to have a class to represent this concept in Gammapy, the `NDDataArray` but it was removed some time ago. Should we re-introduce it to support more cases? 

This would be basically the `IRF` base class without the specific methods and functionality required for IRFs. IRF would inherit from it and re-implement their specific features. It would be a `Quantity` array with associated `MapAxes` and interpolation, get_by_coord evaluation/integral methods.

Opinions @adonath , @maxnoe , @kosack ?

**Comment by adonath** [comment]
@registerrier Yes, we could certainly abstract away some more generalized functionality from the `IRF` class. However I think from the Gammapy side this would only make real sense if we re-use it then for `Map` as well. Otherwise we do not gain anything in terms of code-duplication, because the `NDDataArray` would only be used for `IRF` otherwise. 

What we would need is a `NDDataArray` class, that explicitly allows combinations such as `[wcs_axes, energy_axis]`, or `[hpx_axis, energy_axis]`, which I think is not covered by the CTAO data model and is also not required from their side.

So right now I don't see any advantage in doing this, except if it happens in an even more generalized way to handle WCS and HEALPix axes as well.

**Comment by adonath** [comment]
Or more specifically: we would require an `NDDataArray` object, that can handle "correlated" coordinates. I.e. coordinates which cannot be represented as an outer product. I can't remember whether there was any specification about this in the CTAO document.
